From 96b57fdd283c3856f390e971d3acbe3f88107c0d Mon Sep 17 00:00:00 2001
From: Sowmiya Sree Elavalagan <quic_ssreeela@quicinc.com>
Date: Tue, 15 Mar 2022 13:49:02 -0700
Subject: [PATCH] mbssid: support extended rates in non-tx profiles

Add extended supported rates for non-transmitting interface if
applicable. If the transmitting interface supports it and
the non-transmitting interface does not then add it in the
non-inheritance element.

Signed-off-by: Sowmiya Sree Elavalagan <quic_ssreeela@quicinc.com>
---
 src/ap/ap_config.h  |  1 +
 src/ap/ieee802_11.c | 11 +++++++++++
 2 files changed, 12 insertions(+)

diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index b519992e212a..ebcca548d8cd 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -911,6 +911,7 @@ struct hostapd_bss_config {
 
 	u8 rnr;
 	int beacon_tx_mode;
+	bool xrates_supported;
 };
 
 /**
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index cb7107af69fe..0df209e0b157 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -166,6 +166,7 @@ u8 * hostapd_eid_ext_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	int i, num, count;
 	int h2e_required;
 
+	hapd->conf->xrates_supported = 0;
 	if (hapd->iface->current_rates == NULL)
 		return eid;
 
@@ -215,6 +216,7 @@ u8 * hostapd_eid_ext_supp_rates(struct hostapd_data *hapd, u8 *eid)
 			*pos++ = 0x80 | BSS_MEMBERSHIP_SELECTOR_SAE_H2E_ONLY;
 	}
 
+	hapd->conf->xrates_supported = 1;
 	return pos;
 }
 
@@ -7774,6 +7776,10 @@ static size_t hostapd_eid_mbssid_elem_len(struct hostapd_data *hapd,
 			ie_count++;
 		if (!rsnx && hostapd_wpa_ie(tx_bss, WLAN_EID_RSNX))
 			ie_count++;
+		if (bss->conf->xrates_supported)
+			nontx_profile_len += 8;
+		else if (hapd->conf->xrates_supported)
+			ie_count++;
 		if (ie_count)
 			nontx_profile_len += (4 + ie_count);
 
@@ -7925,6 +7931,11 @@ static u8 * hostapd_eid_mbssid_elem(struct hostapd_data *hapd, u8 *eid, u8 *end,
 			non_inherit_ie[ie_count] = WLAN_EID_RSNX;
 			ie_count++;
 		}
+		if (hapd->conf->xrates_supported &&
+		    !bss->conf->xrates_supported) {
+			non_inherit_ie[ie_count] = WLAN_EID_EXT_SUPP_RATES;
+			ie_count++;
+		}
 		if (ie_count) {
 			*eid++ = WLAN_EID_EXTENSION;
 			*eid++ = 2 + ie_count;
-- 
2.31.1

